tools/nemo_forced_aligner/utils/data_prep.py

```diff
@@ -137,157 +244,356 @@ def get_y_and_boundary_info_for_utt(text, model, separator):
-                }
+                for token_i, (token, token_id, token_cased) in enumerate(
+                    zip(word_tokens, word_token_ids, word_tokens_cased)
+                ):
+                    # add the text tokens and the blanks in between them
+                    # to our token-based variables
+                    utt.token_ids_with_blanks.extend([token_id, BLANK_ID])
+                    utt.segments_and_tokens[-1].words_and_tokens[-1].tokens.append(
+                        Token(
+                            text=token,
+                            text_cased=token_cased,
+                            s_start=len(utt.token_ids_with_blanks) - 2,
+                            s_end=len(utt.token_ids_with_blanks) - 2,
+                        )
+                    )
+
+                    if token_i < len(word_tokens) - 1:
+                        utt.segments_and_tokens[-1].words_and_tokens[-1].tokens.append(
+                            Token(
+                                text=BLANK_TOKEN,
+                                text_cased=BLANK_TOKEN,
+                                s_start=len(utt.token_ids_with_blanks) - 1,
+                                s_end=len(utt.token_ids_with_blanks) - 1,
+                            )
+                        )
+
+                if word_i < len(words) - 1:
+                    utt.segments_and_tokens[-1].words_and_tokens.append(
+                        Token(
+                            text=BLANK_TOKEN,
+                            text_cased=BLANK_TOKEN,
+                            s_start=len(utt.token_ids_with_blanks) - 1,
+                            s_end=len(utt.token_ids_with_blanks) - 1,
+                        )
+                    )
+
+            # add the blank token in between segments/after the final segment
+            utt.segments_and_tokens.append(
+                Token(
+                    text=BLANK_TOKEN,
+                    text_cased=BLANK_TOKEN,
+                    s_start=len(utt.token_ids_with_blanks) - 1,
+                    s_end=len(utt.token_ids_with_blanks) - 1,
+                )
             )
-            segment_s_pointer += len(segment_tokens) * 2
 
-        return y_token_ids_with_blanks, token_info, word_info, segment_info
+        utt.S = len(utt.token_ids_with_blanks)
+
+        return utt
 
     elif hasattr(model.decoder, "vocabulary"):  # i.e. tokenization is simply character-based
 
         BLANK_ID = len(model.decoder.vocabulary)  # TODO: check this is correct
         SPACE_ID = model.decoder.vocabulary.index(" ")
 
-        y_token_ids_with_blanks = [BLANK_ID]
-        token_info = [{"text": BLANK_TOKEN, "s_start": 0, "s_end": 0,}]
-        word_info = []
-        segment_info = []
+        utt.token_ids_with_blanks = [BLANK_ID]
+
+        # check for text being 0 length
+        if len(text) == 0:
+            utt.S = len(utt.token_ids_with_blanks)
```

Comment:
Avoid single character variable names

Reply:
Good point, I am going to just get rid of the variable since whenever I need to access that I can just call `len(utt.token_ids_with_blanks)` instead of `utt.S`